modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,470 @@
+ * }
+ */
+export const timeoutCache = {};
+
+/**
+ * Cleanup timer for the adIdLookup and timeoutCache caches. If all works properly then the caches are self-cleaning
+ * but in case something goes sideways we poll periodically to cleanup old values to prevent a memory leak
+ */
+let cacheCleanupInterval;
+
+let conversantAnalytics = Object.assign(
+  adapter({URL, ANALYTICS_TYPE}),
+  {
+    track({eventType, args}) {
+      if (DO_SAMPLE) {
+        utils.logMessage(eventType, Object.assign({}, args));
+        switch (eventType) {
+          case AUCTION_END:
+            onAuctionEnd(args);
+            break;
+          case AD_RENDER_FAILED:
+            onAdRenderFailed(args);
+            break;
+          case BID_WON:
+            onBidWon(args);
+            break;
+          case BID_TIMEOUT:
+            onBidTimeout(args);
+            break;
+        } // END switch
+      } // END IF(doSample)
+    } // END track()
+  }
+);
+
+// ================================================== EVENT HANDLERS ===================================================
+/**
+ * We get the list of timeouts before the endAution, cache them temporarily in a global cache and the endAuction event
+ * will pick them up.  Uses getLookupKey() to create the key to the entry from auctionId, adUnitCode and bidderCode.
+ * Saves a single value of timeReceived so we can do cache purging periodically.
+ *
+ * Current assumption is that the timeout will always be an array even if it is just one object in the array.
+ * @param args  [{
+  "bidId": "80882409358b8a8",
+    "bidder": "conversant",
+    "adUnitCode": "MedRect",
+    "auctionId": "afbd6e0b-e45b-46ab-87bf-c0bac0cb8881"
+  }, {
+    "bidId": "9da4c107a6f24c8",
+    "bidder": "conversant",
+    "adUnitCode": "Leaderboard",
+    "auctionId": "afbd6e0b-e45b-46ab-87bf-c0bac0cb8881"
+  }
+ ]
+ */
+function onBidTimeout(args) {
+  args.forEach(timedOutBid => {
+    const timeoutCacheKey = getLookupKey(timedOutBid.auctionId, timedOutBid.adUnitCode, timedOutBid.bidder);
+    timeoutCache[timeoutCacheKey] = {
+      timeReceived: Date.now()
+    }
+  });
+}
+
+/**
+ * Bid won occurs after auctionEnd so we need to send this separately. We also save an entry in the adIdLookup cache
+ * so that if the render fails we can match up important data so we can send a valid RENDER FAILED event back.
+ * @param args bidWon args
+ */
+function onBidWon(args) {
+  const bidderCode = args.bidderCode;
+  const adUnitCode = args.adUnitCode;
+  const auctionId = args.auctionId;
+  // Make sure we have all the data we need
+  if (typeof bidderCode === 'undefined' || typeof adUnitCode === 'undefined' || typeof auctionId === 'undefined') {
+    utils.logError(LOG_PREFIX + 'onBidWon() did not get all the necessary data to process the event.');
+    return;
+  }
+  const bidWonPayload = createPayload('bid_won', auctionId);
+
+  const adUnitPayload = createAdUnit();
+  bidWonPayload.adUnits[adUnitCode] = adUnitPayload;
```

Comment:
Can there be more than one ad unit in a bidWon event?

Reply:
no, bid won is specific to one adunit for one bidder